tls/s2n_signature_scheme.c

```diff
@@ -194,55 +198,73 @@ const struct s2n_signature_scheme* const s2n_supported_sig_scheme_pref_list[] =
-        /* RSA PSS - Commented out until Handshake can negotiate it */
-        /* &s2n_rsa_pss_pss_sha256, */
-        /* &s2n_rsa_pss_pss_sha384, */
-        /* &s2n_rsa_pss_pss_sha512, */
-        /* &s2n_rsa_pss_rsae_sha256, */
-        /* &s2n_rsa_pss_rsae_sha384, */
-        /* &s2n_rsa_pss_rsae_sha512, */
+/* The original preference list, but with rsa_pss supported. */
+const struct s2n_signature_scheme* const s2n_sig_scheme_pref_list_20200207[] = {
+        /* RSA PSS */
+        &s2n_rsa_pss_pss_sha256,
+        &s2n_rsa_pss_pss_sha384,
+        &s2n_rsa_pss_pss_sha512,
+        &s2n_rsa_pss_rsae_sha256,
+        &s2n_rsa_pss_rsae_sha384,
+        &s2n_rsa_pss_rsae_sha512,
 
         /* RSA PKCS1 */
         &s2n_rsa_pkcs1_sha256,
         &s2n_rsa_pkcs1_sha384,
         &s2n_rsa_pkcs1_sha512,
         &s2n_rsa_pkcs1_sha224,
-        &s2n_rsa_pkcs1_sha1,
 
-        /* ECDSA */
-        &s2n_ecdsa_sha256,
-        &s2n_ecdsa_sha384,
+        /* ECDSA - TLS 1.2 */
+        &s2n_ecdsa_sha256, /* same iana value as TLS 1.3 s2n_ecdsa_secp256r1_sha256 */
+        &s2n_ecdsa_secp256r1_sha256,
+        &s2n_ecdsa_sha384, /* same iana value as TLS 1.3 s2n_ecdsa_secp384r1_sha384 */
+        &s2n_ecdsa_secp384r1_sha384,
         &s2n_ecdsa_sha512,
         &s2n_ecdsa_sha224,
+
+        /* SHA-1 Legacy */
+        &s2n_rsa_pkcs1_sha1,
         &s2n_ecdsa_sha1,
 };
 
-/* Signature Scheme Preference List:  TLS 1.3 */
-/* This list MUST NOT contain any ECDSA s2n_signature_scheme's with a NULL signature_curve (except ECDSA_SHA1). */
-const struct s2n_signature_scheme * const s2n_tls13_sig_scheme_pref_list[] = {
-        /* RSA PSS - Commented out until it is actually supported */
-        /* &s2n_rsa_pss_rsae_sha256, */
-        /* &s2n_rsa_pss_rsae_sha384, */
-        /* &s2n_rsa_pss_rsae_sha512, */
-
-        /* RSA PKCS1 */
-        &s2n_rsa_pkcs1_sha256,
-        &s2n_rsa_pkcs1_sha384,
-        &s2n_rsa_pkcs1_sha512,
+const struct s2n_signature_preferences s2n_signature_preferences_original = {
+        .count = s2n_array_len(s2n_sig_scheme_pref_list_original),
+        .signature_schemes = s2n_sig_scheme_pref_list_original,
+};
 
-        /* ECDSA */
-        &s2n_ecdsa_secp256r1_sha256,
-        &s2n_ecdsa_secp384r1_sha384,
+const struct s2n_signature_preferences s2n_signature_preferences_20200207 = {
+        .count = s2n_array_len(s2n_sig_scheme_pref_list_20200207),
+        .signature_schemes = s2n_sig_scheme_pref_list_20200207,
+};
 
-        /* SHA-1 Legacy */
-        &s2n_rsa_pkcs1_sha1,
-        &s2n_ecdsa_sha1,
+static struct {
+    const char *version;
+    const struct s2n_signature_preferences *preferences;
+} selection[] = {
+        {.version = "default", .preferences = &s2n_signature_preferences_original },
+        {.version = "20200207", .preferences = &s2n_signature_preferences_20200207 },
+        {.version = "original", .preferences = &s2n_signature_preferences_original },
```

Comment:
can we have a better name?

Reply:
How about "oldest" or "oldest_supported"? Or "legacy"?

Reply:
"pre-20200207"? :)

Reply:
`"20140601"` :)